includes/class-user.php

```diff
@@ -0,0 +1,212 @@
+		// Generate a token.
+		$token = wp_generate_password( 32, false );
+
+		// Save the token.
+		update_user_meta( $user_id, self::TECNAVIA_TOKEN_META_FIELD, $token );
+
+		// Save the token expiration time.
+		update_user_meta( $user_id, self::TECNAVIA_TOKEN_META_FIELD . '_expiration_time', time() + self::TOKEN_EXPIRATION_TIME );
+
+		// Return the token.
+		return $token;
+	}
+
+	/**
+	 * Gets the user token.
+	 *
+	 * @param int $user_id The user ID.
+	 * @return string
+	 */
+	public static function get_user_token( $user_id ) {
+		$token = get_user_meta( $user_id, self::TECNAVIA_TOKEN_META_FIELD, true );
+
+		$token = self::maybe_refresh_token( $token, $user_id );
+
+		return $token;
+	}
+
+	/**
+	 * Maybe refreshes the user token.
+	 *
+	 * @param string $token The user token.
+	 * @param int    $user_id The user ID.
+	 * @return string
+	 */
+	public static function maybe_refresh_token( $token, $user_id ) {
+		if ( empty( $token ) || self::is_user_token_expired( $user_id ) ) {
+			$token = self::create_user_token( $user_id );
+		}
+
+		return $token;
+	}
+
+	/**
+	 * Checks if the user token is expired.
+	 *
+	 * @param int $user_id The user ID.
+	 * @return bool
+	 */
+	public static function is_user_token_expired( $user_id ) {
+		// Get the token expiration time.
+		$token_expiration_time = get_user_meta( $user_id, self::TECNAVIA_TOKEN_META_FIELD . '_expiration_time', true );
+
+		return $token_expiration_time < time();
+	}
+
+	/**
+	 * Gets the user Tecnavia URL.
+	 *
+	 * @param int $user_id The user ID.
+	 * @return string
+	 */
+	public static function get_user_tecnavia_url( $user_id ) {
+		// Get the user token.
+		$token = self::get_user_token( $user_id );
+
+		// Get the Tecnavia URL.
+		$tecnavia_url = Settings::get_tecnavia_url();
+
+		// Check if the Tecnavia URL is empty.
+		if ( empty( $tecnavia_url ) || empty( $token ) ) {
+			return '';
+		}
+
+		// Generate user's e-edition URL.
+		$user_tecnavia_url = add_query_arg( sanitize_key( 'token' ), sanitize_text_field( $token ), $tecnavia_url );
+
+		return $user_tecnavia_url;
+	}
+
+	/**
+	 * Checks if the user has access to Tecnavia.
+	 *
+	 * @param int $user_id The user ID.
+	 *
+	 * @return bool
+	 */
+	public static function check_user_tecnavia_access( $user_id ) {
+		// Get the tecnavia access permissions.
+		$tecnavia_access_permissions = Settings::get_tecnavia_access_permissions();
+
+		/**
+		 * 1. Check if all registered users have access.
+		 */
+		$all_registered_users_have_access = 'yes' === $tecnavia_access_permissions[ Settings::ALL_REGISTERED_USERS_ACCESS_KEY ] ? true : false;
+
+		// Since all registered users have access, return true.
+		if ( $all_registered_users_have_access ) {
+			return true;
+		}
+
+		/**
+		 * 2. Check if the user's role has access.
+		 */
+		$allowed_roles = $tecnavia_access_permissions[ Settings::ALLOWED_ROLES_ACCESS_KEY ];
+		$user_object   = get_userdata( $user_id );
+		$user_role     = array_shift( $user_object->roles );
```

Comment:
It's common for some publishers to use plugins that allow users to have multiple roles. Let's check the entire `roles` array instead of doing `array_shift`